### PR TITLE
Music: advanced functions

### PR DESCRIPTION
**Pull request by breville**
This adds full Blockly function support to the advanced model in Music Lab, for internal evaluation by the curriculum team.  

Here's an example demonstrating `when run`, triggers, and functions that support variable numbers of parameters and optional return values:
Screenshot 2024-09-05 at 12 12 24 PM

And here is the functions category in the toobox:
Screenshot 2024-09-05 at 12 12 57 PM

### Details

While the `simple2` model continues to use `Blockly.JavaScript.blockToCode` for `when run` and each trigger, to get a separate piece of executable JS for each such "entrypoint", the `advanced` model now uses `Blockly.JavaScript.workspaceToCode` for each such entrypoint, and adds an internal context variable to each so that the JS code knows what it should run.  By using `Blockly.JavaScript.workspaceToCode`, each function's JS is generated properly and is available to all entrypoints.

Here's an illustrative example of the JS generated for a trigger entrypoint:

```
var __context = "trigger1";
var currentTime, offset;

function play_at_offset(offset) {
  Sequencer.playSoundAtMeasureById("electro/synth_cave", offset, false, "82{1N}kn{he?^ze9x$|~");
}

if (__context == 'when_run') {
  Sequencer.playSoundAtMeasureById("electro/whua", 1, true, "P6;omww?#*zzN?m|IJgd");
}

currentTime = startPosition;
if (__context == "trigger1") {
  play_at_offset(9);
}
```

I did look into behaving more like Sprite Lab and and having the block generator for a trigger register a callback.  However, it appears that for `CustomMarshalingInterpreter.createNativeFunctionFromInterpreterFunction` to be available, we need to initialize `JSInterpreter` which in turn has a dependency on `studioApp`.  So I opted for the above approach instead.

### CODE REVIEW

File: apps/src/music/blockly/setup.ts [hunk 1]
```diff
@@ -52,18 +54,22 @@ export function setUpBlocklyForMusicLab() {
...
-  Blockly.cdoUtils.registerCustomProcedureBlocks();
+  if (getBlockMode() !== BlockMode.ADVANCED) {
+    // Override default function block implementation.
+    Blockly.cdoUtils.registerCustomProcedureBlocks();
+
+    // Remove two default entries in the toolbox's Functions category that
+    // we don't want.
+    delete Blockly.Blocks.procedures_defreturn;
+    delete Blockly.Blocks.procedures_ifreturn;
+  }
```
Comment (mikeharv) [inline apps/src/music/blockly/setup.ts:65]: This is the approach I was thinking about too. It feels like a good compromise to give us a fresh starting place for this advanced context while keeping the other custom blocks untouched for other labs (and simple music lab).

Sort of unrelated, but I imagine that at some point we'd want `blockMode` to be a level config setting too?

File: apps/src/music/blockly/toolbox.js [hunk 2]
```diff
@@ -448,6 +448,7 @@ export function getToolbox(toolbox) {
         },
         {
           includeVariables: true,
+          includeFunctions: true,
```
Comment (mikeharv) [inline apps/src/music/blockly/toolbox.js:451]: I also expected this change - looks good.

For my understanding, are these generated toolboxes just defaults for each block mode? In other words, if a writer specifies a toolbox configuration in the level data json, does it override this one?

**Review by mikeharv — APPROVED** [review]
I know less about the need for internal context variables and the entrypoints mentioned in the description, but the changes to block definitions and toolboxes look appropriate to me.
I think it should be stressed that these blocks are primarily meant to unblock internal evaluation. I think there's still a few open questions around UI as well as whether we want to have a separate block for parameters (intended to be used within the scope of a function's definition stack) and variables (global, can be used anywhere). Since either of these questions could lead to differences in the final workspace serialization, it might be advisable to avoid using them in levels that we care about not breaking.

File: apps/src/music/blockly/setup.ts (hunk 1 above)
Comment (breville) [reply apps/src/music/blockly/setup.ts:65]: > Sort of unrelated, but I imagine that at some point we'd want blockMode to be a level config setting too?

Good question.  Yes, it seems likely we'll need to do that fairly soon

File: apps/src/music/blockly/toolbox.js (hunk 2 above)
Comment (breville) [reply apps/src/music/blockly/toolbox.js:451]: > For my understanding, are these generated toolboxes just defaults for each block mode?

Yes, we just use the default toolboxes for variables and functions, handled [here](https://github.com/code-dot-org/code-dot-org/blob/63de45fc6dd2ace65ab4914454e6d102e036cfa2/apps/src/music/blockly/toolbox.js#L366-L385).  

>  In other words, if a writer specifies a toolbox configuration in the level data json, does it override this one?

In Music Lab, the optional toolbox configuration is used to specify a subset of entries to be shown, such as [here](https://github.com/code-dot-org/code-dot-org/blob/63de45fc6dd2ace65ab4914454e6d102e036cfa2/dashboard/config/levels/custom/music/Music%20Level%202.level#L15-L22).

Comment (mikeharv) [reply apps/src/music/blockly/toolbox.js:451]: So the level configuration entries must be included in the generated toolbox for the given block mode, and it specifies that we should only show the blocks indicated by those entries?

Comment (breville) [reply apps/src/music/blockly/toolbox.js:451]: The level configuration is optional.  If not specified, then the entire set of blocks for that block mode (i.e. `simple2` or `advanced`) is shown.  If specified, then only the entries included are shown.

Comment (mikeharv) [reply apps/src/music/blockly/toolbox.js:451]: Got it!

File: apps/src/music/blockly/MusicBlocklyWorkspace.ts [hunk 3]
```diff
@@ -255,7 +258,8 @@ export default class MusicBlocklyWorkspace {
         const id = block.getFieldValue(TRIGGER_FIELD);
         this.compiledEvents[triggerIdToEvent(id)] = {
           code:
-            Blockly.JavaScript.blockToCode(block) + functionImplementationsCode,
+            `var __context = "${id}";\n` +
+            Blockly.JavaScript.workspaceToCode(this.workspace),
```
Comment (breville) [inline apps/src/music/blockly/MusicBlocklyWorkspace.ts:262]: My bad: this change regressed `simple2` behavior, causing all workspace code to execute when any trigger was handled, causing users to see the results of all triggers, not just the one invoked.  This is fixed with https://github.com/code-dot-org/code-dot-org/pull/61105.

**Pull request merged**